[RFC] tools: add options to optimize host binaries

rockdrilla opened the pull request:
Mains goals are:
- reduce binary size of host tools;
- reduce i/o load on build host;
- increase performance of host tools being built.

rockdrilla commented:
@Ansuel, @blogic, @hauke, @nbd168, friendly ping and PTAL :)

rockdrilla commented:
I've encountered bug with `tools/make-ext4fs`, MR with fix: openwrt/make_ext4fs#3

Ansuel commented:
@rockdrilla lovely! I wonder if you make some stats about size and time difference for this.

Also why RFC? The changes are so simple I guess this can be merged right away. Maybe the submenu and using "Optimize tool" might be confusing. I don't remember what we use for kernel build option but we should stick to that naming scheme.

Maybe a better name for the category might be "Advanced Host build Options" ?

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,61 @@ menuconfig DEVEL
 		  universal precompiled host tools archive to use in another buildroot.
 
+	menuconfig OPTIMIZE_HOST_TOOLS
+		bool "Optimize host tools" if DEVEL
```
lets use "Host Tools Compile Options"

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,61 @@ menuconfig DEVEL
+	if OPTIMIZE_HOST_TOOLS
+
+		config HOST_FLAGS_OPT
+			string "Generic host optimization flags"
```
"Host Tools optimization flags"

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,61 @@ menuconfig DEVEL
+		comment "Miscellaneous flags"
+
+		config HOST_EXTRA_CFLAGS
+			string "Extra host CFLAGS"
```
Host Tools, same for the other 2.

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,61 @@ menuconfig DEVEL
+			  in order to reduce binary size.
+
+		config HOST_FLAGS_STRIP
+			string "Linker strip flags"
```
Host Tool Linker ... ?

rockdrilla commented:
@Ansuel I've changed titles / help messages, PTAL =)

rockdrilla commented:
Now it looks fine from my point of view:

![изображение](https://github.com/user-attachments/assets/eb0e7f7f-2d1c-4cdf-a425-68a93f766e12)

Ansuel commented on rules.mk:
```diff
@@ -278,12 +278,20 @@ PKG_CONFIG:=$(STAGING_DIR_HOST)/bin/pkg-config
 
 export PKG_CONFIG
 
+HOST_FLAGS_OPT:=$(if $(CONFIG_OPTIMIZE_HOST_TOOLS),$(call qstrip,$(CONFIG_HOST_FLAGS_OPT)))
+HOST_FLAGS_OPT:=$(if $(HOST_FLAGS_OPT),$(HOST_FLAGS_OPT),-O2)
```
Pack this under a single condition.

rockdrilla replied:
It's not possible due to promise "`If not set or empty, defaults to "-O2".`"

Ansuel replied:
`HOST_FLAGS_OPT:=$(if $(CONFIG_OPTIMIZE_HOST_TOOLS),$(call qstrip,$(CONFIG_HOST_FLAGS_OPT)),-O2)` ?

rockdrilla replied:
Seems legit.

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,68 @@ menuconfig DEVEL
+	menuconfig OPTIMIZE_HOST_TOOLS
+		bool "Host Tools Compile Options" if DEVEL
+
+	if OPTIMIZE_HOST_TOOLS
```
Indent everything of one additional block since it's part of the menuconfig.

Ansuel commented on config/Config-devel.in:
```diff
@@ -62,6 +62,68 @@ menuconfig DEVEL
+
+		config HOST_FLAGS_OPT
+			string "Host Tools optimization flags"
+			depends on OPTIMIZE_HOST_TOOLS
```
drop depends on OPTIMIZE_HOST_TOOLS as we already have the if for everything.

rockdrilla replied:
Clause "`depends on OPTIMIZE_HOST_TOOLS`" ensures that options from this submenu doesn't appear in config if `CONFIG_OPTIMIZE_HOST_TOOLS` is not set. Same for `CONFIG_HOST_TOOLS_STRIP` and `CONFIG_HOST_FLAGS_STRIP`.

Ansuel replied:
Nono I verified. The "if OPTIMIZE_HOST_TOOLS" is enough to make the option not appear in the final .config

Indeed from the kernel kconfig documentation is precisely said that either if CONFIG block or depends on CONFIG should be used as they do the same exact thing. (I just verified that to make sure)

rockdrilla committed:
tools: add options to optimize host binaries

Mains goals are:
- reduce binary size of host tools;
- reduce i/o load on build host;
- increase performance of host tools being built.

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18659
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
